### PR TITLE
Implement SAFE_HEAP using acorn, and use it on user JS

### PR DESCRIPTION
**Pull request by kripken**
Previously we instrumented the compiled code, and also code in JS libraries
using `{{{ makeGetValue }}}` etc., but not arbitrary `HEAP8` etc. usage in user
JS code. This fixed that by adding a `safeHeap` pass to the acorn optimizer.

This found a few issues in our JS code too, which have been fixed in
recent PRs already.

### CODE REVIEW

File: tests/test_other.py [hunk 1]
```diff
@@ -1716,7 +1716,7 @@ def test_extern_prepost(self):
       self.assertGreater(len(js), 100 * SLACK)
 
   def test_js_optimizer(self):
-    ACORN_PASSES = ['JSDCE', 'AJSDCE', 'applyImportAndExportNameChanges', 'emitDCEGraph', 'applyDCEGraphRemovals', 'growableHeap', 'unsignPointers', 'asanify']
+    ACORN_PASSES = ['JSDCE', 'AJSDCE', 'applyImportAndExportNameChanges', 'emitDCEGraph', 'applyDCEGraphRemovals', 'growableHeap', 'unsignPointers', 'asanify', 'safeHeap']
```
Comment (sbc100) [inline tests/test_other.py:1719]: Time to split this one one-element-per-line, or at least multiple lines?

**Review by sbc100 — APPROVED** [review]
Do we still need the JS library technique of using `{{{ makeGetValue }}}` or can that now be removed in favor of this?

**Comment by kripken** [comment]
> Do we still need the JS library technique of using `{{{ makeGetValue }}}` or can that now be removed in favor of this?

We can maybe remove it, yeah. There is *some* advantage to the makeGetValue notation, in that you don't need to write the `>> 2` etc., which can be error prone. But I agree it's probably not worth it given the verbosity, overall it's less clear. If you agree I can look into it.

**Comment by sbc100** [comment]
> > Do we still need the JS library technique of using `{{{ makeGetValue }}}` or can that now be removed in favor of this?
> 
> We can maybe remove it, yeah. There is _some_ advantage to the makeGetValue notation, in that you don't need to write the `>> 2` etc., which can be error prone. But I agree it's probably not worth it given the verbosity, overall it's less clear. If you agree I can look into it.

I wasn't suggesting removing `{{{ makeGetValue }}}` just that it should not try do any SAFE_HEAP stuff.. then the SAFE_HEAP stuff is all in one place and will be consistent between user code and library code?

**Comment by kripken** [comment]
Oh, yes, it is redundant atm. Rather than remove it, though, I'd prefer to eventually remove `{{{ makeGetValue }}}` entirely.

**Pull request merged**